src/main/java/com/teragrep/cfe_39/consumers/kafka/queue/WritableQueue.java

```diff
@@ -0,0 +1,144 @@
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.function.BiPredicate;
+import java.util.function.ToLongFunction;
+import java.util.stream.Stream;
+
+public class WritableQueue {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(WritableQueue.class);
+
+    private final Path queueDirectory;
+    private String queueNamePrefix;
+
+    public WritableQueue(String queueDirectory) {
+        this.queueDirectory = Paths.get(queueDirectory);
+        this.queueNamePrefix = "";
+        if (!Files.isDirectory(this.queueDirectory)) {
+            throw new IllegalArgumentException("Provided path is not a " + "directory <[" + queueDirectory + "]>");
+        }
+        if (!Files.isWritable(this.queueDirectory)) {
+            throw new IllegalArgumentException("Provided path is not " + "writeable <[" + queueDirectory + "]>");
+        }
+    }
+
+    private File getNextWritableFilename() throws IOException {
+
+        try (
+                Stream<Path> files = Files.find(queueDirectory, 1, getFileMatcher(queueNamePrefix), FileVisitOption.FOLLOW_LINKS)
+        ) {
+
+            long sequenceNumber = files.mapToLong(getPathToSequenceNumberFunction()).max().orElse(0);
+
+            long nextSequenceNumber = sequenceNumber + 1;
+
+            // create next
+            return new File(
+                    queueDirectory.toAbsolutePath() + File.separator + queueNamePrefix + "." + nextSequenceNumber
+            );
+        }
+        catch (UncheckedIOException uncheckedIOException) {
+            // just retry, reader modified the directory
+            return getNextWritableFilename();
+        }
+    }
+
+    public File getNextWritableFile() throws IOException {
+        if (queueNamePrefix.isEmpty()) {
+            throw new IOException("No queueNamePrefix set");
+        }
+        else {
+            return getNextWritableFilename();
+        }
+    }
+
+    public void setQueueNamePrefix(String a) {
+        this.queueNamePrefix = a;
+    }
+
+    private BiPredicate<Path, BasicFileAttributes> getFileMatcher(String queueNamePrefix) {
+        return (path, basicFileAttributes) -> {
+            if (!path.getFileName().toString().startsWith(queueNamePrefix)) {
+                return false;
+            }
+            else if (path.getFileName().toString().endsWith(".state")) {
+                return false;
+            }
+            else if (!basicFileAttributes.isRegularFile()) {
+                return false;
+            }
+            else {
+                LOGGER.trace("getFileMatcher returning: <{}>", path);
+                return true;
+            }
+        };
+    }
+
+    private ToLongFunction<Path> getPathToSequenceNumberFunction() {
+        return path -> {
+            String pathString = path.toString();
+
+            int dotPosition = pathString.lastIndexOf('.');
+
+            String sequenceNumberString = pathString.substring(dotPosition + 1);
```

Comment:
Limited to just single character long sequences?

Reply:
pathString.substring(dotPosition + 1) extracts the whole substring after the dotPosition, thus sequenceNumberString can be longer than a single character.